File: shared/web-sdk/src/connectors/task-status-connector.ts
import type { TaskStatus } from "@sentient/protocol";
import type { Connector, SentientSDKInternal } from "../connector-types.ts";

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------

export interface TaskSnapshotItem {
  readonly taskId: string;
  readonly toolName: string;
  readonly cycleId: string;
  readonly status: TaskStatus;
  /** Auto-derived short preview of the tool's args. */
  readonly argsPreview: string;
  readonly startedAtMs: number;
  readonly endedAtMs?: number;
}

export interface TaskStatusConnectorConfig {
  /** Called on every individual update. */
  onUpdate?: (item: TaskSnapshotItem) => void;
  /** Called whenever the task list changes — ordered by startedAt ascending. */
  onList?: (items: readonly TaskSnapshotItem[]) => void;
}

// ---------------------------------------------------------------------------
// TaskStatusConnector — session-scoped live task list.
//
// Capability: "task.status"
// Direction: status (observer; no per-task outbound protocol from here)
//
// Receives:
//   - task.update  (one on register, one on deregister; dedups by taskId)
//
// Exposes ordered list by startedAt ascending. Terminal states stay in
// the list — UI can filter out if it only wants running tasks.
// ---------------------------------------------------------------------------

export class TaskStatusConnector implements Connector {
  readonly capability = "task.status";
  readonly kind = "status" as const;

  private readonly config: TaskStatusConnectorConfig;
  private unsubs: (() => void)[] = [];
  private tasks = new Map<string, TaskSnapshotItem>();

  constructor(config: TaskStatusConnectorConfig = {}) {
    this.config = config;
  }

  list(): readonly TaskSnapshotItem[] {
    return [...this.tasks.values()].sort((a, b) => a.startedAtMs - b.startedAtMs);
  }

  attach(sdk: SentientSDKInternal): void {
    this.tasks = new Map();

    this.unsubs.push(
      sdk.onMessage("task.update", (msg: unknown) => {
        const m = msg as {
          taskId?: string;
          toolName?: string;
          cycleId?: string;
          status?: TaskStatus;
          argsPreview?: string;
          startedAtMs?: number;
          endedAtMs?: number;
        };
        if (!m.taskId || !m.toolName || !m.cycleId || !m.status || m.startedAtMs === undefined) return;
        const item: TaskSnapshotItem = {
          taskId: m.taskId,
          toolName: m.toolName,
          cycleId: m.cycleId,
          status: m.status,
          argsPreview: m.argsPreview ?? "",
          startedAtMs: m.startedAtMs,
          ...(m.endedAtMs !== undefined ? { endedAtMs: m.endedAtMs } : {}),
        };
        this.tasks.set(item.taskId, item);
        this.config.onUpdate?.(item);
        this.config.onList?.(this.list());
      }),
    );
  }

  detach(): void {
    for (const unsub of this.unsubs) unsub();
    this.unsubs = [];
    this.tasks = new Map();
  }
}

File: shared/mobile-sdk/src/commonMain/kotlin/io/sentient/mobilesdk/connectors/TaskStatusConnector.kt
// ---------------------------------------------------------------------------
// TaskStatusConnector — session-scoped live tool-call list.
//
// Mirrors web-sdk's task-status-connector.ts:
//   capability = "task.status"  (status observer; no outbound protocol)
//
//   turn.tool.update → upsert by toolCallId into a map. A malformed frame
//                      decodes to ServerMessage.Unknown and never reaches here,
//                      so the TS "reject if a required field is missing" guard is
//                      satisfied by the wire schema. Fire onUpdate(item) + onList(list()).
//
//   list() → all rows sorted by startedAtMs ASCENDING. Terminal states stay in
//            the list — the UI filters running-only if it wants. endedAtMs is
//            carried through once the call reaches a terminal status.
//
// Threading: single-threaded; the orchestrator routes frames + subscribes the
// callbacks. The mutable map is owned here.
// ---------------------------------------------------------------------------
package io.sentient.mobilesdk.connectors

import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.protocol.SdkEvent
import io.sentient.mobilesdk.protocol.ServerMessage

/**
 * Immutable snapshot of one tool-call row. Built from [ServerMessage.TurnToolUpdate].
 *
 * IDENTITY IS [toolCallId] — one row per model-emitted tool call. [taskId] is present
 * only for a BACKGROUND tool (delegateTask), which returns a handle immediately and
 * completes later via a stimulus; a foreground tool has none.
 */
data class TaskSnapshotItem(
    val toolCallId: String,
    val toolName: String,
    val turnId: String,
    /** "running" | "done" | "error". */
    val status: String,
    /** Auto-derived short preview of the tool's args; "" when absent. USER CONTENT — never logged. */
    val argsPreview: String,
    val startedAtMs: Long,
    val endedAtMs: Long? = null,
    val taskId: String? = null,
)

class TaskStatusConnector(
    private val onUpdate: ((TaskSnapshotItem) -> Unit)? = null,
    private val onList: ((List<TaskSnapshotItem>) -> Unit)? = null,
    private val onEvent: ((SdkEvent) -> Unit)? = null,
) : Connector {
    override val capability: String = CAPABILITY

    private val log = createLogger("connector", "task-status")

    private val tasks = LinkedHashMap<String, TaskSnapshotItem>()

    /** All rows, ordered by startedAtMs ascending. Safe to read synchronously. */
    fun list(): List<TaskSnapshotItem> = tasks.values.sortedBy { it.startedAtMs }

    override fun handle(msg: ServerMessage) {
        when (msg) {
            is ServerMessage.TurnToolUpdate -> onToolUpdate(msg)
            else -> Unit // not owned by this connector
        }
    }

    private fun onToolUpdate(msg: ServerMessage.TurnToolUpdate) {
        val item = TaskSnapshotItem(
            toolCallId = msg.toolCallId,
            toolName = msg.toolName,
            turnId = msg.turnId,
            status = msg.status,
            argsPreview = msg.argsPreview,
            startedAtMs = msg.startedAtMs,
            endedAtMs = msg.endedAtMs,
            taskId = msg.taskId,
        )
        // argsPreview is user content — ids/status/timing only.
        log.info(
            "turn.tool.update",
            mapOf(
                "toolCallId" to item.toolCallId,
                "turnId" to item.turnId,
                "toolName" to item.toolName,
                "status" to item.status,
                "taskId" to item.taskId,
                "endedAtMs" to item.endedAtMs,
            ),
        )
        tasks[item.toolCallId] = item
        onEvent?.invoke(SdkEvent.TaskUpserted(item))
        onUpdate?.invoke(item)
        onList?.invoke(list())
    }

    /** Clear the row list. Called by the orchestrator on session teardown. */
    fun clear() {
        log.info("clear", mapOf("count" to tasks.size))
        tasks.clear()
    }

    companion object {
        const val CAPABILITY: String = "task.status"
    }
}
